presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -462,116 +366,265 @@ else if (!result.getBlocked().isDone()) {
 
     private boolean isReadyForExecution(StreamingPlanSection section)
     {
-        SqlStageExecution stageExecution = getStageExecution(section.getPlan().getFragment().getId());
-        if (stageExecution.getState() != PLANNED) {
+        Optional<SectionExecution> sectionExecution = getLatestSectionExecution(getStageId(section.getPlan().getFragment().getId()));
+        if (sectionExecution.isPresent() && (sectionExecution.get().isRunning() || sectionExecution.get().isFinished())) {
```

Comment:
Will there be any surprise when `SectionExecution` change there state during this method get called? :) (e.g. from wait -> running, or from running -> finished/failed)

Reply:
I don't think so.  This is the same logic as before, but we can't just check != PLANNED because we want to include FAILED tasks for retries, so we check for running and successfully finished stages instead.